addons/payment/controllers/portal.py

```diff
@@ -99,8 +99,9 @@ def payment_pay(
         # Instantiate transaction values to their default if not set in parameters
         reference = reference or payment_utils.singularize_reference_prefix(prefix='tx')
         amount = amount or 0.0  # If the amount is invalid, set it to 0 to stop the payment flow
-        company_id = company_id or partner_sudo.company_id.id or user_sudo.company_id.id
-        currency_id = currency_id or request.env['res.company'].browse(company_id).currency_id.id
+        res_company = company_id and request.env['res.company'].sudo().browse(company_id) or partner_sudo.company_id or user_sudo.company_id
+        company_id = res_company.id
+        currency_id = currency_id or res_company.currency_id.id
```

Comment:
I'd find it clearer if you kept the way `company_id` is evaluated untouched. When mixing `and`, `or`'s `browse`, etc., the logic doesn't stand out anymore. Thanks to the prefetch, we don't need to worry about the possibly redundant browse.

Also, `res_company` can be simply named `company`. It's not necessary to follow the naming of the model if it doesn't help with code clarity.

```suggestion
        company_id = company_id or partner_sudo.company_id.id or user_sudo.company_id.id
        company = request.env['res.company'].browse(company_id)
        currency_id = currency_id or company.currency_id.id
```